packages/backend/src/engine/amplify_stack.ts

```diff
@@ -20,3 +30,66 @@ export class AmplifyStack extends Stack {
     return defaultId;
   };
 }
+
+const cognitoRoleTrustPolicyValidator: IAspect = {
+  visit: (node: IConstruct) => {
+    if (!(node instanceof Role)) {
+      return;
+    }
+    const assumeRolePolicyDocument = node.assumeRolePolicy?.toJSON();
+    if (!assumeRolePolicyDocument) {
+      return;
+    }
+
+    assumeRolePolicyDocument.Statement.forEach(
+      cognitoTrustPolicyStatementValidator
+    );
+  },
+};
+
+const cognitoTrustPolicyStatementValidator = ({
+  Action: action,
+  Condition: condition,
+  Effect: effect,
+  Principal: principal,
+}: {
+  // These property names come from the IAM policy document which we do not control
+  /* eslint-disable @typescript-eslint/naming-convention */
+  Action: string;
+  Condition?: Record<string, Record<string, string>>;
+  Effect: 'Allow' | 'Deny';
+  Principal?: { Federated?: string };
+  /* eslint-enable @typescript-eslint/naming-convention */
+}) => {
+  if (action !== 'sts:AssumeRoleWithWebIdentity') {
+    return;
+  }
+  if (principal?.Federated !== 'cognito-identity.amazonaws.com') {
+    return;
+  }
+  if (effect === 'Deny') {
+    return;
+  }
+  // if we got here, we have a policy that allows AssumeRoleWithWebIdentity with Cognito
+  // need to validate that the policy has an appropriate condition
+
+  const audCondition =
+    condition?.StringEquals?.['cognito-identity.amazonaws.com:aud'];
+  if (typeof audCondition !== 'string' || audCondition.length === 0) {
+    throw new AmplifyFault('InvalidTrustPolicyFault', {
+      message:
+        'Cannot create a Role trust policy with Cognito that does not have a StringEquals condition for cognito-identity.amazonaws.com:aud',
+    });
+  }
+
+  const amrCondition =
+    condition?.['ForAnyValue:StringLike']?.[
+      'cognito-identity.amazonaws.com:amr'
+    ];
+  if (typeof amrCondition !== 'string' || amrCondition.length === 0) {
+    throw new AmplifyFault('InvalidTrustPolicyFault', {
```

Comment:
this should be fault for resources we create but error if it's customer defined policy.

Reply:
While true, I don't think we have enought context here to make that distinction. I'd rather just classify as a fault and reevaluate if that proves to be noisy.